File: tools/js/eat_parser.ts
// eat_parser.ts
// EAT → JSON parser and JSON → EAT serializer

export type EatValue = string | boolean | string[] | Record<string, any> | Record<string, any>[];
export interface EatDocument { [key: string]: EatValue; }

interface ParseState {
  currentBlock: string | null;
  currentKeys: string[];
  inMultiline: boolean;
  multilineBuffer: string[];
  inArray: boolean;
}

const arrayHeaderRegex = /^(\w+)\[(\d*)\]\{([^}]*)\}:\s*$/;
const blockHeaderRegex = /^(\w+)(?:\{([^}]*)\})?:\s*$/;
const exactHeaderRegex = /^(\w+_exact):\s*$/;

export function parseEAT(text: string): EatDocument {
  const lines = text.split(/\r?\n/);
  const doc: EatDocument = {};
  const state: ParseState = {
    currentBlock: null,
    currentKeys: [],
    inMultiline: false,
    multilineBuffer: [],
    inArray: false,
  };

  for (let rawLine of lines) {
    const line = rawLine.replace(/\s+$/, "");
    const trimmed = line.trim();

    if (!trimmed && !state.inMultiline) {
      if (state.currentBlock && state.currentKeys.length === 0 && !state.inArray) {
        state.currentBlock = null;
      }
      continue;
    }

    if (state.inMultiline && trimmed === '"""') {
      if (state.currentBlock) {
        doc[state.currentBlock] = state.multilineBuffer.join("\n");
      }
      state.inMultiline = false;
      state.currentBlock = null;
      state.multilineBuffer = [];
      continue;
    }

    if (state.inMultiline) {
      state.multilineBuffer.push(line);
      continue;
    }

    if (trimmed.endsWith('"""')) {
      const name = trimmed.split(":")[0];
      state.currentBlock = name;
      state.inMultiline = true;
      state.multilineBuffer = [];
      continue;
    }

    const arrayMatch = trimmed.match(arrayHeaderRegex);
    if (arrayMatch) {
      const [, name, , keysRaw] = arrayMatch;
      state.currentBlock = name;
      state.currentKeys = keysRaw.split(",").map((s) => s.trim());
      state.inArray = true;
      doc[name] = [];
      continue;
    }

    if (state.inArray && state.currentBlock && state.currentKeys.length > 0 && trimmed.includes(",")) {
      const values = trimmed.split(",").map((s) => s.trim());
      const row: Record<string, any> = {};
      state.currentKeys.forEach((k, i) => {
        row[k] = values[i] ?? null;
      });
      (doc[state.currentBlock] as any[]).push(row);
      continue;
    }

    const inlineMatch = trimmed.match(/^(\w+):\s+(.+)$/);
    if (inlineMatch) {
      const [, name, rawValue] = inlineMatch;
      let value: any = rawValue.trim();
      if (value === "true") {
        value = true;
      } else if (value === "false") {
        value = false;
      } else if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
        value = value.slice(1, -1);
      }
      doc[name] = value;
      state.currentBlock = null;
      state.currentKeys = [];
      state.inArray = false;
      continue;
    }

    const exactMatch = trimmed.match(exactHeaderRegex);
    if (exactMatch) {
      state.currentBlock = exactMatch[1];
      state.currentKeys = [];
      state.inArray = false;
      continue;
    }

    if (state.currentBlock && state.currentBlock.endsWith("_exact")) {
      const value = trimmed.replace(/^"|"$/g, "");
      doc[state.currentBlock] = value;
      state.currentBlock = null;
      state.currentKeys = [];
      state.inArray = false;
      continue;
    }

    const blockMatch = trimmed.match(blockHeaderRegex);
    if (blockMatch) {
      const [, name, keysRaw] = blockMatch;
      state.currentBlock = name;
      state.currentKeys = keysRaw ? keysRaw.split(",").map((s) => s.trim()) : [];
      state.inArray = false;

      if (!keysRaw) {
        doc[name] = [];
      }
      continue;
    }

    if (state.currentBlock && state.currentKeys.length > 0 && !state.inArray) {
      const values = trimmed.split(",").map((s) => s.trim());
      const obj: Record<string, any> = {};
      state.currentKeys.forEach((k, i) => {
        obj[k] = values[i] ?? null;
      });
      doc[state.currentBlock] = obj;
      state.currentBlock = null;
      state.currentKeys = [];
      continue;
    }

    if (state.currentBlock && state.currentKeys.length === 0 && !state.inArray) {
      const existing = doc[state.currentBlock];
      const val = trimmed.replace(/,$/, "");
      if (Array.isArray(existing)) {
        existing.push(val);
      } else if (existing === undefined) {
        doc[state.currentBlock] = [val];
      } else {
        doc[state.currentBlock] = [existing, val];
      }
      continue;
    }
  }

  return doc;
}

export function stringifyEAT(doc: EatDocument): string {
  const lines: string[] = [];

  for (const [key, value] of Object.entries(doc)) {
    if (key.endsWith("_exact") && typeof value === "string") {
      lines.push(`${key}:`);
      lines.push(`  "${value}"`);
      lines.push("");
      continue;
    }

    if (Array.isArray(value)) {
      if (value.length === 0) {
        lines.push(`${key}:`);
        lines.push("");
        continue;
      }
      if (typeof value[0] === "string") {
        lines.push(`${key}:`);
        for (const v of value as string[]) {
          lines.push(`  ${v}`);
        }
        lines.push("");
        continue;
      }
      const arr = value as Record<string, any>[];
      const keys = Object.keys(arr[0] || {});
      lines.push(`${key}[${arr.length}]{${keys.join(", ")}}:`);
      for (const row of arr) {
        const vals = keys.map((k) => String(row[k] ?? ""));
        lines.push(`  ${vals.join(", ")}`);
      }
      lines.push("");
      continue;
    }

    if (typeof value === "object" && value !== null) {
      const obj = value as Record<string, any>;
      const keys = Object.keys(obj);
      lines.push(`${key}{${keys.join(", ")}}:`);
      const vals = keys.map((k) => String(obj[k] ?? ""));
      lines.push(`  ${vals.join(", ")}`);
      lines.push("");
      continue;
    }

    if (typeof value === "string") {
      lines.push(`${key}:`);
      lines.push(`  ${value}`);
      lines.push("");
      continue;
    }

    if (typeof value === "boolean") {
      lines.push(`${key}: ${value ? "true" : "false"}`);
      lines.push("");
      continue;
    }
  }

  return lines.join("\n").trimEnd();
}

File: tools/py/eat_parser.py
"""eat_parser.py
Simple EAT → dict parser and dict → EAT serializer in Python.
"""
import re
from typing import Any, Dict, List, Union

EatValue = Union[str, bool, List[str], Dict[str, Any], List[Dict[str, Any]]]
EatDocument = Dict[str, EatValue]

array_header_re = re.compile(r'^(\w+)\[(\d*)\]\{([^}]*)\}:\s*$')
block_header_re = re.compile(r'^(\w+)(?:\{([^}]*)\})?:\s*$')
exact_header_re = re.compile(r'^(\w+_exact):\s*$')
inline_kv_re = re.compile(r'^(\w+):\s+(.+)$')


def parse_eat(text: str) -> EatDocument:
    lines = text.splitlines()
    doc: EatDocument = {}

    current_block: str | None = None
    current_keys: List[str] = []
    in_multiline = False
    multiline_buffer: List[str] = []
    in_array = False

    for raw_line in lines:
        line = raw_line.rstrip()
        trimmed = line.strip()

        if not trimmed and not in_multiline:
            if current_block and not current_keys and not in_array:
                current_block = None
            continue

        if in_multiline and trimmed == '"""':
            if current_block:
                doc[current_block] = "\n".join(multiline_buffer)
            in_multiline = False
            current_block = None
            multiline_buffer = []
            continue

        if in_multiline:
            multiline_buffer.append(line)
            continue

        if trimmed.endswith('"""'):
            name = trimmed.split(":", 1)[0]
            current_block = name
            in_multiline = True
            multiline_buffer = []
            continue

        m_array = array_header_re.match(trimmed)
        if m_array:
            name, _, keys_raw = m_array.groups()
            current_block = name
            current_keys = [k.strip() for k in keys_raw.split(",")]
            in_array = True
            doc[name] = []
            continue

        if in_array and current_block and current_keys and "," in trimmed:
            values = [v.strip() for v in trimmed.split(",")]
            row = {k: (values[i] if i < len(values) else None) for i, k in enumerate(current_keys)}
            assert isinstance(doc[current_block], list)
            doc[current_block].append(row)
            continue

        m_inline = inline_kv_re.match(trimmed)
        if m_inline:
            name, raw_value = m_inline.groups()
            value: Any = raw_value.strip()
            lower = value.lower()
            if lower == "true":
                value = True
            elif lower == "false":
                value = False
            elif (value.startswith('"') and value.endswith('"')) or (value.startswith("'") and value.endswith("'")):
                value = value[1:-1]
            doc[name] = value
            current_block = None
            current_keys = []
            in_array = False
            continue

        m_exact = exact_header_re.match(trimmed)
        if m_exact:
            current_block = m_exact.group(1)
            current_keys = []
            in_array = False
            continue

        if current_block and current_block.endswith("_exact"):
            value = trimmed.strip('"')
            doc[current_block] = value
            current_block = None
            current_keys = []
            in_array = False
            continue

        m_block = block_header_re.match(trimmed)
        if m_block:
            name, keys_raw = m_block.groups()
            current_block = name
            current_keys = [k.strip() for k in keys_raw.split(",")] if keys_raw else []
            in_array = False
            if not keys_raw:
                doc[name] = []
            continue

        if current_block and current_keys and not in_array:
            values = [v.strip() for v in trimmed.split(",")]
            obj = {k: (values[i] if i < len(values) else None) for i, k in enumerate(current_keys)}
            doc[current_block] = obj
            current_block = None
            current_keys = []
            continue

        if current_block and not current_keys and not in_array:
            existing = doc.get(current_block)
            val = trimmed.rstrip(",")
            if existing is None:
                doc[current_block] = [val]
            elif isinstance(existing, list):
                existing.append(val)
            else:
                doc[current_block] = [existing, val]
            continue

    return doc


def stringify_eat(doc: EatDocument) -> str:
    lines: List[str] = []
    for key, value in doc.items():
        if key.endswith("_exact") and isinstance(value, str):
            lines.append(f"{key}:")
            lines.append(f'  "{value}"')
            lines.append("")
            continue

        if isinstance(value, list):
            if not value:
                lines.append(f"{key}:")
                lines.append("")
                continue

            if isinstance(value[0], str):
                lines.append(f"{key}:")
                for v in value:
                    lines.append(f"  {v}")
                lines.append("")
                continue

            if isinstance(value[0], dict):
                keys = list(value[0].keys())
                lines.append(f"{key}[{len(value)}]" + "{" + ", ".join(keys) + "}:")
                for row in value:
                    vals = [str(row.get(k, "")) for k in keys]
                    lines.append("  " + ", ".join(vals))
                lines.append("")
                continue

        if isinstance(value, dict):
            keys = list(value.keys())
            lines.append(f"{key}" + "{" + ", ".join(keys) + "}:")
            vals = [str(value.get(k, "")) for k in keys]
            lines.append("  " + ", ".join(vals))
            lines.append("")
            continue

        if isinstance(value, str):
            lines.append(f"{key}:")
            lines.append(f"  {value}")
            lines.append("")
            continue

        if isinstance(value, bool):
            lines.append(f"{key}: {'true' if value else 'false'}")
            lines.append("")
            continue

    return "\n".join(lines).rstrip() + "\n"
